crates/proc_macro_srv/src/proc_macro/mod.rs

```diff
@@ -77,7 +77,7 @@ impl FromStr for TokenStream {
 /// with `Delimiter::None` delimiters and negative numeric literals.
 impl fmt::Display for TokenStream {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str(&self.to_string())
+        f.write_str(&self.0.to_string())
```

Comment:
I recommend that do not changes this file as it is almost straight copy from rustc such that it is quite hard to trace if they changed something relevant.

Reply:
I understand, however without this change, afaict the Display impls are recursive because the blanket ToString impl calls the Display impl a'la https://doc.rust-lang.org/stable/src/alloc/string.rs.html#2194-2207

Reply:
Yes, but these won't be called in normal situation and these implementation are just dummy for make the ABI happy.

Anyway, I think it is okay to change it for functionality

Reply:
My fear is that if they're there for ABI compatibility then it implies they could be called at which point they'd break.  If on the other hand they can't be called then removing them (commenting them out?) might be a better bet since it'd be clearer what was going on.  I'm not clear about how that might affect compatibility though, you're the expert here :D

Reply:
Okay, Agree. Maybe add a comment to indicate it is a complicated case :)

> I'm not clear about how that might affect compatibility though, you're the expert here :D

I wish I were the expert. :( I even don't understand why the original rustc allow duplication implementation of the `ToString` It should be conflicted to `impl<T: fmt::Display + ?Sized> ToString` implementation :

https://github.com/rust-lang/rust/blob/b33e234155b33ab6bce280fb2445b62b68622b61/library/proc_macro/src/lib.rs#L138-L155

Reply:
I think that when the real code is built, specialisation is possible.  I'll consider a suitable comment to add